### PR TITLE
Add a value wrapper that preserves cache metadata

### PR DESCRIPTION
**Pull request by ohthehugemanatee**
Adds a new class to wrap around resolved values while preserving cache metadata. Applies it to the built in Entity resolvers.

You can build your value in one line with 
``` new CacheableGraphqlValue($value, [$list, $of, $cache, $dependencies])```
Or modify an existing one with setValue(), or any of the CacheableMetadata methods.

### CODE REVIEW

File: src/GraphQL/CacheableGraphqlValue.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+ * Wrapper class for values resolved through GraphQL resolvers, which includes
+ * Drupal cache metadata.
+ */
+
+namespace Drupal\graphql\GraphQL;
+
+use Drupal\Core\Cache\CacheableDependencyInterface;
+use Drupal\Core\Cache\CacheableMetadata;
+
+class CacheableGraphqlValue extends CacheableMetadata {
```
Comment (fubhy) [inline src/GraphQL/CacheableGraphqlValue.php]: Let's call this CacheableLeafValue. It should only be used for leaf values and that makes it clearer.

File: src/GraphQL/CacheableGraphqlValue.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
+<?php
+/**
+ * @file
```
Comment (fubhy) [inline src/GraphQL/CacheableGraphqlValue.php]: That should be the class documentation header and not (at)file. Let's put that above the class instead.

File: src/GraphQL/CacheableGraphqlValue.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+
+namespace Drupal\graphql\GraphQL;
+
+use Drupal\Core\Cache\CacheableDependencyInterface;
+use Drupal\Core\Cache\CacheableMetadata;
+
+class CacheableGraphqlValue extends CacheableMetadata {
+
+  /**
+   * @var mixed The actual value being wrapped.
```
Comment (fubhy) [inline src/GraphQL/CacheableGraphqlValue.php]: Can you break the description string into its own line like we do in other places conforming to the Drupal coding standards?

File: src/GraphQL/CacheableGraphqlValue.php [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+   */
+  protected $value;
+
+  /**
+   * CacheableGraphqlValue constructor.
+   *
+   * @param mixed $value The actual value to be wrapped.
```
Comment (fubhy) [inline src/GraphQL/CacheableGraphqlValue.php]: Can you break the description string into its own line like we do in other places conforming to the Drupal coding standards?

File: src/GraphQL/CacheableGraphqlValue.php [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+        $this->addCacheableDependency($dependency);
+      }
+    }
+  }
+
+  public function setValue($value) {
+    $this->value = $value;
+  }
+
+  public function getValue() {
```
Comment (fubhy) [inline src/GraphQL/CacheableGraphqlValue.php]: Misses docblock.

File: src/GraphQL/CacheableGraphqlValue.php [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+  public function __construct($value, $dependencies = []) {
+    $this->setValue($value);
+    foreach ($dependencies as $dependency) {
+      if ($dependency instanceof CacheableDependencyInterface) {
+        $this->addCacheableDependency($dependency);
+      }
+    }
+  }
+
+  public function setValue($value) {
```
Comment (fubhy) [inline src/GraphQL/CacheableGraphqlValue.php]: Misses docblock.

File: src/GraphQL/Execution/Processor.php [hunk 7]
```diff
@@ -59,6 +60,11 @@ protected function doResolve(FieldInterface $field, AstFieldInterface $ast, $par
       $this->metadata->setCacheMaxAge(0);
     }
 
+    // If it's a graphQL cacheable value wrapper, extract the real value to return.
```
Comment (fubhy) [inline src/GraphQL/Execution/Processor.php]: GraphQL

File: src/GraphQL/Field/FieldBase.php [hunk 8]
```diff
@@ -32,4 +34,15 @@ public function getType() {
...
   }
+
+  /**
+   * Resolve function for this field.
+   * @param $value
+   * @param array $args
+   * @param \Youshido\GraphQL\Execution\ResolveInfo $info
+   * @return \Drupal\graphql\GraphQL\CacheableGraphqlValue
+   */
+  public function resolve($value, array $args, ResolveInfo $info) {
```
Comment (fubhy) [inline src/GraphQL/Field/FieldBase.php]: Let's not add this.

**Comment by ohthehugemanatee** [comment]
Agree with every one of the suggestions, and fixed them all. :)

**Comment by fubhy** [comment]
@ohthehugemanatee Can you submit another PR that adds https://github.com/fubhy/graphql-drupal/blob/8.x-3.x/src/GraphQL/CacheableLeafValue.php for the examples module?

**Pull request merged**